fast_lio_sam_qn/src/fast_lio_sam_qn.cpp

```diff
@@ -0,0 +1,467 @@
+        gtsam_graph_.add(gtsam::BetweenFactor<gtsam::Pose3>(current_keyframe_idx_-1, current_keyframe_idx_, pose_from.between(pose_to), odom_noise));
+        init_esti_.insert(current_keyframe_idx_, pose_to);
+      }
+      current_keyframe_idx_++;
+
+      //// 3. vis
+      high_resolution_clock::time_point t3 = high_resolution_clock::now();
+      {
+        std::lock_guard<std::mutex> lock(vis_mutex_);
+        updateOdomsAndPaths(current_frame_);
+      }
+
+      //// 4. optimize with graph
+      high_resolution_clock::time_point t4 = high_resolution_clock::now();
+      // m_corrected_esti = gtsam::LevenbergMarquardtOptimizer(m_gtsam_graph, init_esti_).optimize(); // cf. isam.update vs values.LM.optimize
+      {
+        std::lock_guard<std::mutex> lock(graph_mutex_);
+        isam_handler_->update(gtsam_graph_, init_esti_);
+        isam_handler_->update();
+        if (loop_added_flag_) //https://github.com/TixiaoShan/LIO-SAM/issues/5#issuecomment-653752936
+        {
+          isam_handler_->update();
+          isam_handler_->update();
+          isam_handler_->update();
+        }
+        gtsam_graph_.resize(0);
+        init_esti_.clear();
+      }
+
+      //// 5. handle corrected results
+      // get corrected poses and reset odom delta (for realtime pose pub)
+      high_resolution_clock::time_point t5 = high_resolution_clock::now();
+      {
+        std::lock_guard<std::mutex> lock(realtime_pose_mutex_);
+        corrected_esti_ = isam_handler_->calculateEstimate();
+        last_corrected_pose_ = gtsamPoseToPoseEig(corrected_esti_.at<gtsam::Pose3>(corrected_esti_.size()-1));
+        odom_delta_ = Eigen::Matrix4d::Identity();
+      }
+      // correct poses in keyframes
+      if (loop_added_flag_)
+      {
+        std::lock_guard<std::mutex> lock(keyframes_mutex_);
+        for (size_t i = 0; i < corrected_esti_.size(); ++i)
+        {
+          keyframes_[i].pose_corrected_eig_ = gtsamPoseToPoseEig(corrected_esti_.at<gtsam::Pose3>(i));
+        }
+        loop_added_flag_ = false;
+      }
+      high_resolution_clock::time_point t6 = high_resolution_clock::now();
+
+      ROS_INFO("real: %.1f, key_add: %.1f, vis: %.1f, opt: %.1f, res: %.1f, tot: %.1fms", 
+        duration_cast<microseconds>(t2 - t1).count()/1e3, duration_cast<microseconds>(t3 - t2).count()/1e3,
+        duration_cast<microseconds>(t4 - t3).count()/1e3, duration_cast<microseconds>(t5 - t4).count()/1e3,
+        duration_cast<microseconds>(t6 - t5).count()/1e3, duration_cast<microseconds>(t6 - t1).count()/1e3);
+    }
+  }
+  return;
+}
+
+void FastLioSamQn::loopTimerFunc(const ros::TimerEvent& event)
+{
+  auto &latest_keyframe = keyframes_.back();
+  if (!is_initialized_ || keyframes_.empty() || latest_keyframe.processed_) return;
+  latest_keyframe.processed_ = true;
+
+  high_resolution_clock::time_point t1 = high_resolution_clock::now();
+  const int closest_keyframe_idx = loop_closure_->fetchClosestKeyframeIdx(latest_keyframe, keyframes_);
```

Comment:
Why do we use `const int` here?

Reply:
You mean why didn't I use `const int &`? In case of primitive types, there are little performance difference with and without &.